core/src/main/scala/org/apache/spark/Partition.scala

```diff
@@ -26,6 +26,11 @@ trait Partition extends Serializable {
    */
   def index: Int
 
+  override def equals(other: Any): Boolean = other match {
+    case that: Partition => super.equals(that) && that.index == index
```

Comment:
If all the subclasses override these methods (and some implement some custom logic), then this isn't used, and maybe it's simpler to omit it. If this stays, yes, you're right that it really has to check the class of itself vs the argument too.